src/main/java/org/candlepin/subscriptions/marketplace/MarketplacePayloadMapper.java

```diff
@@ -0,0 +1,199 @@
+    }
+
+    /**
+     * Create UsageRequest pojo to send to Marketplace
+     *
+     * @param tallySummary TallySummary
+     * @return UsageRequest
+     */
+    public UsageRequest createUsageRequest(TallySummary tallySummary) {
+
+        UsageRequest usageRequest = new UsageRequest();
+
+        var usageEvents = produceUsageEvents(tallySummary);
+        usageEvents.forEach(usageRequest::addDataItem);
+
+        log.debug("UsageRequest {}", usageRequest);
+
+        return usageRequest;
+    }
+
+    /**
+     * We only want to send snapshot information for OpenShift-metrics, OpenShift-dedicated-metrics product
+     * ids.  To prevent duplicate data, we don't want to send snapshots with the Usage or ServiceLevel of
+     * "_ANY".  We only want to report on hourly metrics, so the Granularity should be HOURLY.
+     *
+     * @param snapshot tally snapshot
+     * @return eligibility status
+     */
+    protected boolean isSnapshotPAYGEligible(TallySnapshot snapshot) {
+        String productId = snapshot.getProductId();
+        var applicableProducts = List.of("OpenShift-metrics", "OpenShift-dedicated-metrics");
+        boolean isApplicableProduct = applicableProducts.contains(productId);
+
+        boolean isHourlyGranularity = Objects.nonNull(snapshot.getGranularity()) &&
+            Objects.equals(TallySnapshot.Granularity.HOURLY, snapshot.getGranularity());
+
+        boolean isSpecificUsage = Objects.nonNull(snapshot.getUsage()) &&
+            !Objects.equals(snapshot.getUsage(), TallySnapshot.Usage.ANY);
+
+        boolean isSpecificServiceLevel =
+            Objects.nonNull(snapshot.getSla()) && !Objects.equals(snapshot.getSla(), TallySnapshot.Sla.ANY);
+
+        boolean isSnapshotPAYGEligible =
+            isHourlyGranularity && isApplicableProduct && isSpecificUsage && isSpecificServiceLevel;
+
+        if (!isSnapshotPAYGEligible) {
+            log.debug("Snapshot not eligible for sending to RHM {}", snapshot);
+        }
+        return isSnapshotPAYGEligible;
+    }
+
+    /**
+     * UsageRequest objects are made up of a list of UsageEvents.
+     *
+     * @param tallySummary TallySummary
+     * @return List<UsageEvent>
+     */
+    protected List<UsageEvent> produceUsageEvents(TallySummary tallySummary) {
+        if (Objects.isNull(tallySummary.getTallySnapshots())) {
+            tallySummary.setTallySnapshots(new ArrayList<>());
+        }
+
+        return tallySummary.getTallySnapshots().stream().filter(this::isSnapshotPAYGEligible)
+            .map(snapshot -> {
+                String productId = snapshot.getProductId();
+                OffsetDateTime snapshotDate = snapshot.getSnapshotDate();
+                String eventId = snapshot.getId().toString();
+
+                // call MarketplaceIdProvider.findSubscriptionId once available
+                UsageCalculation.Key usageKey = new UsageCalculation.Key(productId,
+                    ServiceLevel.fromString(snapshot.getSla().toString()),
```

Comment:
This is funny.  What type is `snapshot.getSla()`? Some generated `.dto.ServiceLevel` class or something?

Reply:
Ugh yessssssss, it's generated from a file in the `schema` folder